cpp/src/parquet/file_reader.cc

```diff
@@ -366,13 +366,19 @@ class SerializedFile : public ParquetFileReader::Contents {
         std::make_shared<::arrow::io::internal::ReadRangeCache>(source_, ctx, options);
     std::vector<::arrow::io::ReadRange> ranges;
     prebuffered_column_chunks_.clear();
+    int num_cols = file_metadata_->num_columns();
+    // a bitmap for buffered columns.
+    std::shared_ptr<Buffer> buffer_columns;
+    if (!row_groups.empty()) {
```

Comment:
a possible optimization is that we can regard `nullptr` as buffer the column.

```c++
    if (!row_groups.empty()) {
      if (num_cols == static_cast<int>(row_groups.size())) {
        buffer_columns = nullptr;
      } else {
        PARQUET_THROW_NOT_OK(AllocateEmptyBitmap(num_cols, properties_.memory_pool())
                                 .Value(&buffer_columns));
        for (int col : column_indices) {
          ::arrow::bit_util::SetBit(buffer_columns->mutable_data(), col);
        }
      }
    }
```

And if 

```
  void PreBuffer(const std::vector<int>& row_groups,
                 const std::vector<int>& column_indices,
                 const ::arrow::io::IOContext& ctx,
                 const ::arrow::io::CacheOptions& options)
```

Find that the `buffer` is nullptr, it will just buffer the column.

Reply:
Using `buffer=nullptr` as a placeholder for marking all columns are pre-buffered seems ok. Then we should not add a row group to the map if the none of columns in the row group are pre-buffered.

Besides, if only a single column is not pre-buffered in the row group, the overhead is still high to set bit for other columns.

Reply:
@pitrou hi pitrou, what do you think of this? If it's ok I can have a try
Or we can run a benchmark first to check that is this can benefits our regression?

Reply:
AFAIU, the regression is not Parquet-related.

Reply:
Oh, I see. So lets me keep this. If I found there are remain performance issues, I will add `buffer=nullptr` check